dvc/repo/ls.py

```diff
@@ -0,0 +1,85 @@
+import os
+
+from dvc.exceptions import PathMissingError, OutputNotFoundError
+
+
+@staticmethod
+def ls(url, target=None, rev=None, recursive=None, outs_only=False):
+    from dvc.external_repo import external_repo
+    from dvc.repo import Repo
+    from dvc.utils import relpath
+
+    with external_repo(url, rev) as repo:
+        result = []
+        if isinstance(repo, Repo):
+            result.extend(_ls_outs_repo(repo, target, recursive))
+        files = _ls_files_repo(repo, target, recursive)
+
+        if target and not len(result) and not len(files):
+            raise PathMissingError(target, repo)
+
+        if not outs_only:
+            result.extend(files)
+        target_path_info = _get_target_path_info(repo, target)
```

Comment:
I'm pretty sure you can move this up and just pass target_path_info instead of `target` to other methods. Currently you are calling this helper three times for no reason :slightly_smiling_face:

Reply:
> Currently you are calling this helper three times for no reason

The idea was to keep function interface lightweight and simple - `_ls_outs` and `_ls_files` requires only `repo` which is transparent and make the functions more simple to use. With `target_path_info` param I don't like much how functions signature looks